src/segment_nightscout.go

```diff
@@ -0,0 +1,131 @@
+package main
+
+import (
+	"encoding/json"
+)
+
+// segment struct, makes templating easier
+type nightscout struct {
+	props *properties
+	env   environmentInfo
+	// array of nightscoutData (is often just 1, and we will pick the ZEROeth)
+
+	Sgv       int64
+	Direction string
+
+	TrendIcon string
+}
+
+const (
+	// Your complete Nightscout URL and APIKey like this
+	URL               Property = "url"
+	DoubleUpIcon      Property = "doubleup_icon"
+	SingleUpIcon      Property = "singleup_icon"
+	FortyFiveUpIcon   Property = "fortyfiveup_icon"
+	FlatIcon          Property = "flat_icon"
+	FortyFiveDownIcon Property = "fortyfivedown_icon"
+	SingleDownIcon    Property = "singledown_icon"
+	DoubleDownIcon    Property = "doubledown_icon"
+)
+
+type nightscoutData struct {
+	Sgv       int64  `json:"sgv"`
+	Direction string `json:"direction"`
+}
+
+func (ns *nightscout) enabled() bool {
+	data, err := ns.getResult()
+	if err != nil {
+		return false
+	}
+	ns.Sgv = data.Sgv
+	ns.Direction = data.Direction
+
+	ns.TrendIcon = ns.getTrendIcon()
+
+	return true
+}
+
+func (ns *nightscout) getTrendIcon() string {
+	switch ns.Direction {
+	case "DoubleUp":
+		return ns.props.getString(DoubleUpIcon, "↑↑")
+	case "SingleUp":
+		return ns.props.getString(SingleUpIcon, "↑")
+	case "FortyFiveUp":
+		return ns.props.getString(FortyFiveUpIcon, "↗")
+	case "Flat":
+		return ns.props.getString(FlatIcon, "→")
+	case "FortyFiveDown":
+		return ns.props.getString(FortyFiveDownIcon, "↘")
+	case "SingleDown":
+		return ns.props.getString(SingleDownIcon, "↓")
+	case "DoubleDown":
+		return ns.props.getString(DoubleDownIcon, "↓↓")
+	default:
+		return ""
+	}
+}
+
+func (ns *nightscout) string() string {
+	segmentTemplate := ns.props.getString(SegmentTemplate, "{{.Sgv}}")
+	template := &textTemplate{
+		Template: segmentTemplate,
+		Context:  ns,
+		Env:      ns.env,
+	}
+	text, err := template.render()
+	if err != nil {
+		return err.Error()
+	}
+
+	return text
+}
+
+func (ns *nightscout) getResult() (*nightscoutData, error) {
+	url := ns.props.getString(URL, "")
+	//cacheTimeout := int64(ns.props.getInt(CacheTimeout, DefaultCacheTimeout))
+	cacheTimeout := int64(5) //nightscout can only give you data every 5 min, todo: make it configurable?
+	response := &nightscoutData{}
+	if cacheTimeout > 0 {
+		// check if data stored in cache
+		val, found := ns.env.cache().get(url)
+		// we got something from the cache
+		if found {
+			err := json.Unmarshal([]byte(val), response)
+			if err != nil {
+				return nil, err
+			}
+			return response, nil
+		}
+	}
+
+	httpTimeout := ns.props.getInt(HTTPTimeout, DefaultHTTPTimeout)
+
+	body, err := ns.env.doGet(url, httpTimeout)
+	if err != nil {
+		return &nightscoutData{}, err
+	}
+	var arr []*nightscoutData
+	err = json.Unmarshal(body, &arr)
+	if err != nil {
+		return &nightscoutData{}, err
+	}
+
+	firstelement := arr[0]
+	firstData, err := json.Marshal(firstelement)
```

Comment:
to avoid doing this, we're better off storing the entire response (`body`). Only means we need a smell helper function (can be scoped inside this function) to parse the response and return the first entity.

Reply:
I looked at this, and tried a few different ways and I'm not seeing how it adds a lot (makes it harder to read for me) and it only runs every 5 min.